libs/cosmos-sdk/baseapp/baseapp_parallel.go

```diff
@@ -410,411 +389,366 @@ func newExecuteResult(r abci.ResponseDeliverTx, ms sdk.CacheMultiStore, counter
-	mu     sync.RWMutex
-	txReps []*executeResult
+	g.clearResultChan(txIndex)
+	g.reRunChan <- txIndex
 }
 
-func newExecResult() *txResultCollector {
-	return &txResultCollector{
-		mu:     sync.RWMutex{},
-		txReps: make([]*executeResult, 0),
-	}
-}
-
-func (e *txResultCollector) clear() {
-	e.mu.Lock()
-	e.txReps = nil
-	e.mu.Unlock()
-}
-
-func (e *txResultCollector) init(txSize int) {
-	txRepsCap := cap(e.txReps)
-	if e.txReps == nil || txRepsCap < txSize {
-		e.txReps = make([]*executeResult, txSize)
-	} else if txRepsCap >= txSize {
-		e.txReps = e.txReps[0:txSize:txRepsCap]
-		// https://github.com/golang/go/issues/5373
-		for i := range e.txReps {
-			e.txReps[i] = nil
+func (g *groupTask) clearResultChan(rerunIndex int) {
+	for true {
+		next, ok := g.nextTx(rerunIndex) // TODO add currIndex
+		if ok {
+			g.putResult(next, nil)
+		} else {
+			return
 		}
+		rerunIndex = next
 	}
 }
 
-func (e *txResultCollector) putResult(index int, txResult *executeResult) {
-	e.mu.Lock()
-	e.txReps[index] = txResult
-	e.mu.Unlock()
-}
-
-func (e *txResultCollector) getTxResult(index int) *executeResult {
-	e.mu.RLock()
-	defer e.mu.RUnlock()
-	return e.txReps[index]
-}
-
-type parallelTxManager struct {
-	blockGasMeterMu     sync.Mutex
-	haveCosmosTxInBlock bool
-	isAsyncDeliverTx    bool
-	workgroup           *asyncWorkGroup
-
-	extraTxsInfo      []*extraDataForTx
-	txResultCollector *txResultCollector
-
-	groupList     map[int][]int
-	nextTxInGroup map[int]int
-	preTxInGroup  map[int]int
+func (g *groupTask) run() {
 
-	mu          sync.RWMutex
-	cms         sdk.CacheMultiStore
-	blockHeight int64
+	for true {
+		select {
+		case txIndex := <-g.txChan:
+			g.mu.Lock()
+			res := g.taskRun(txIndex)
+			if res.paraMsg.UseCurrentState {
+				g.addMultiCache(g.ms, nil)
+				g.ms = res.ms.CacheMultiStore()
+			} else {
+				if res.ms != nil {
+					res.ms.Write()
+				}
+			}
 
-	txSize    int
-	cc        *conflictCheck
-	currIndex int
-	currTxFee sdk.Coins
+			if len(g.reRunChan) == 0 {
+				g.putResult(int(res.counter), res)
+				if n, ok := g.nextTx(txIndex); ok {
+					g.addTask(n)
+				}
+			}
+			g.mu.Unlock()
+
+		case rerunIndex := <-g.reRunChan:
+			g.clearResultChan(rerunIndex)
+			g.addMultiCache(g.ms, nil)
+			g.ms = nil
+			size := len(g.txChan)
+			for index := 0; index < size; index++ {
+				<-g.txChan
+			}
 
-	blockMultiStores *cacheMultiStoreList
-	chainMultiStores *cacheMultiStoreList
+			if n, ok := g.nextTx(rerunIndex); ok {
+				g.addTask(n)
+			}
+		case <-g.stopChan:
+			return
+		}
+	}
 }
 
 func newParallelTxManager() *parallelTxManager {
 	isAsync := sm.DeliverTxsExecMode(viper.GetInt(sm.FlagDeliverTxsExecMode)) == sm.DeliverTxsExecModeParallel
-	return &parallelTxManager{
+	para := &parallelTxManager{
 		blockGasMeterMu:  sync.Mutex{},
 		isAsyncDeliverTx: isAsync,
-		workgroup:        newAsyncWorkGroup(),
+		resultCh:         make(chan int, maxTxNumberInParallelChan),
+		stop:             make(chan struct{}, 1),
```

Comment:
`resultCh:         make(chan int, maxTxNumberInParallelChan)` maxTxNumberInParallelChan should have actual tx num, prevent the std tx large than maxTxNumberInParallelChan in one block

Reply:
done